File: src/feats/debug/tracer.rs
//! DebugTracer：记录特征 DAG 管道中每个阶段的 I/O 值和异常。
use serde::Serialize;
use std::any::Any;
use std::collections::HashMap;
use std::fs;
use std::io::Write;
use std::path::Path;
use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};

/// 阶段类型
#[derive(Debug, Clone, Serialize, PartialEq)]
#[serde(rename_all = "UPPERCASE")]
pub enum StageType {
    DefaultInit,
    RawOverride,
    Operator,
}

/// 值快照：记录值和类型名（JSON 兼容）
#[derive(Debug, Clone, Serialize)]
pub struct ValueSnapshot {
    #[serde(rename = "v")]
    pub value: serde_json::Value,
    #[serde(rename = "t")]
    pub type_name: String,
}

impl ValueSnapshot {
    pub fn of(val: &crate::feats::ops::Fv) -> Self {
        use crate::feats::ops::Fv;
        match val {
            Fv::Int(i) => ValueSnapshot { value: serde_json::json!(*i), type_name: "int".into() },
            Fv::Float(f) => ValueSnapshot { value: serde_json::json!(*f), type_name: "float".into() },
            Fv::Str(s) => ValueSnapshot { value: serde_json::json!(s.clone()), type_name: "str".into() },
            Fv::StrList(l) => ValueSnapshot { value: serde_json::json!(l), type_name: "list[str]".into() },
            Fv::IntList(l) => ValueSnapshot { value: serde_json::json!(l), type_name: "list[int]".into() },
        }
    }
}

/// 异常记录
#[derive(Debug, Clone, Serialize)]
pub struct Anomaly {
    pub feature: String,
    pub reason: String,
}

/// 单个阶段追踪：记录一个算子的 I/O
#[derive(Debug, Clone, Serialize)]
pub struct StageTrace {
    #[serde(rename = "type")]
    pub stage_type: StageType,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub name: String,
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    pub inputs: HashMap<String, ValueSnapshot>,
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    pub outputs: HashMap<String, ValueSnapshot>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub overridden: Vec<String>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub anomalies: Vec<Anomaly>,
}

impl StageTrace {
    pub fn defaults(outputs: HashMap<String, ValueSnapshot>) -> Self {
        StageTrace {
            stage_type: StageType::DefaultInit,
            name: String::new(),
            inputs: HashMap::new(),
            outputs,
            overridden: vec![],
            anomalies: vec![],
        }
    }
    pub fn overrides(overridden: Vec<String>, outputs: HashMap<String, ValueSnapshot>) -> Self {
        StageTrace {
            stage_type: StageType::RawOverride,
            name: String::new(),
            inputs: HashMap::new(),
            outputs,
            overridden,
            anomalies: vec![],
        }
    }
    pub fn operator(
        name: &str,
        inputs: HashMap<String, ValueSnapshot>,
        outputs: HashMap<String, ValueSnapshot>,
    ) -> Self {
        StageTrace {
            stage_type: StageType::Operator,
            name: name.to_string(),
            inputs,
            outputs,
            overridden: vec![],
            anomalies: vec![],
        }
    }
}

/// 单样本追踪
#[derive(Debug, Clone, Serialize)]
pub struct SampleTrace {
    pub sample: usize,
    pub stages: Vec<StageTrace>,
}

/// 调试配置
pub struct DebugConfig {
    pub max_trace_samples: usize,
    pub output_dir: String,
}

/// 调试追踪器：挂在 FeatureDag 上，在 execute() 各阶段记录 I/O。
pub struct DebugTracer {
    config: DebugConfig,
    traces: Mutex<Vec<SampleTrace>>,
    total_seen: Mutex<usize>,
    current: Mutex<Option<SampleTrace>>,
}

impl DebugTracer {
    pub fn new(config: DebugConfig) -> Self {
        DebugTracer {
            config,
            traces: Mutex::new(Vec::new()),
            total_seen: Mutex::new(0),
            current: Mutex::new(None),
        }
    }

    /// 开始追踪一个样本
    pub fn begin_sample(&self) {
        let mut seen = self.total_seen.lock().unwrap();
        if *seen >= self.config.max_trace_samples {
            *self.current.lock().unwrap() = None;
            *seen += 1;
            return;
        }
        *self.current.lock().unwrap() = Some(SampleTrace {
            sample: *seen,
            stages: vec![],
        });
        *seen += 1;
    }

    /// 记录默认值初始化阶段
    pub fn trace_defaults(&self, context: &HashMap<String, &crate::feats::ops::Fv>) {
        let mut curr = self.current.lock().unwrap();
        if let Some(ref mut t) = *curr {
            let outputs: HashMap<String, ValueSnapshot> = context
                .iter()
                .map(|(k, v)| (k.clone(), ValueSnapshot::of(*v)))
                .collect();
            t.stages.push(StageTrace::defaults(outputs));
        }
    }

    /// 记录 raw_inputs 覆盖阶段
    pub fn trace_overrides(
        &self,
        context: &HashMap<String, &crate::feats::ops::Fv>,
        overridden: Vec<String>,
    ) {
        let mut curr = self.current.lock().unwrap();
        if let Some(ref mut t) = *curr {
            let outputs: HashMap<String, ValueSnapshot> = context
                .iter()
                .map(|(k, v)| (k.clone(), ValueSnapshot::of(*v)))
                .collect();
            t.stages.push(StageTrace::overrides(overridden, outputs));
        }
    }

    /// 记录一个算子执行
    pub fn trace_operator(
        &self,
        op_name: &str,
        input_names: &[String],
        input_vals: &[&crate::feats::ops::Fv],
        output_names: &[String],
        output_val: &crate::feats::ops::Fv,
    ) {
        let mut curr = self.current.lock().unwrap();
        if let Some(ref mut t) = *curr {
            let inputs: HashMap<String, ValueSnapshot> = input_names
                .iter()
                .zip(input_vals.iter())
                .map(|(n, v)| (n.clone(), ValueSnapshot::of(*v)))
                .collect();
            let mut outputs = HashMap::new();
            let out_vs = ValueSnapshot::of(output_val);
            for name in output_names {
                outputs.insert(name.clone(), out_vs.clone());
            }
            t.stages
                .push(StageTrace::operator(op_name, inputs, outputs));
        }
    }

    /// 结束当前样本，存入 traces
    pub fn end_sample(&self) {
        let mut curr = self.current.lock().unwrap();
        if let Some(t) = curr.take() {
            self.traces.lock().unwrap().push(t);
        }
    }

    /// 保存 traces 和 summary 到 output_dir
    pub fn save(&self) {
        let dir = &self.config.output_dir;
        if dir.is_empty() {
            return;
        }
        let _ = fs::create_dir_all(dir);
        let ts = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_micros();

        // JSONL traces
        let tp = Path::new(dir).join(format!("traces_{}.jsonl", ts));
        let traces = self.traces.lock().unwrap();
        if let Ok(mut f) = fs::File::create(&tp) {
            for t in traces.iter() {
                if let Ok(line) = serde_json::to_string(t) {
                    let _ = writeln!(f, "{}", line);
                }
            }
        }

        // Summary JSON
        let sp = Path::new(dir).join(format!("summary_{}.json", ts));
        let summary = serde_json::json!({
            "total_samples": traces.len(),
        });
        if let Ok(mut f) = fs::File::create(&sp) {
            let _ = writeln!(
                f,
                "{}",
                serde_json::to_string_pretty(&summary).unwrap_or_default()
            );
        }

        println!("[Debug] summary -> {}", sp.display());
        println!(
            "[Debug] traces -> {} ({} samples)",
            tp.display(),
            traces.len()
        );
    }
}

File: python/src/train/debug/tracer.py
from __future__ import annotations

"""Debug tracer: 逐阶段记录特征预处理管道的输入/输出值和异常。"""
import json
import math
import time
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
from typing import Any


class StageType(Enum):
    DEFAULT_INIT = "DEFAULT_INIT"
    RAW_OVERRIDE = "RAW_OVERRIDE"
    OPERATOR = "OPERATOR"


@dataclass
class DebugConfig:
    feature_filter: str = "all"  # "all" | "embed_only" | ["name1","name2"]
    max_trace_samples: int = 100
    output_dir: str = ""


@dataclass
class ValueSnapshot:
    value: Any
    type_name: str

    @classmethod
    def of(cls, val: Any) -> "ValueSnapshot":
        t = type(val).__name__
        if isinstance(val, list):
            inner = type(val[0]).__name__ if val else "empty"
            t = f"list[{inner}]"
        elif isinstance(val, float) and (math.isnan(val) or math.isinf(val)):
            t = "float(anomaly)"
        return cls(value=val, type_name=t)


@dataclass
class Anomaly:
    feature: str
    reason: str
    value: Any = None


@dataclass
class StageTrace:
    stage_type: StageType
    stage_name: str
    inputs: dict[str, ValueSnapshot] = field(default_factory=dict)
    outputs: dict[str, ValueSnapshot] = field(default_factory=dict)
    overridden: list[str] = field(default_factory=list)
    anomalies: list[Anomaly] = field(default_factory=list)

    def to_dict(self) -> dict:
        d: dict = {"type": self.stage_type.value}
        if self.stage_name:
            d["name"] = self.stage_name
        if self.inputs:
            d["inputs"] = {k: {"v": v.value, "t": v.type_name} for k, v in self.inputs.items()}
        if self.outputs:
            d["outputs"] = {k: {"v": v.value, "t": v.type_name} for k, v in self.outputs.items()}
        if self.overridden:
            d["overridden"] = self.overridden
        if self.anomalies:
            d["anomalies"] = [{"feature": a.feature, "reason": a.reason} for a in self.anomalies]
        return d


class DebugTracer:
    """Tracer attached to FeatureDag — records per-sample stage traces."""

    def __init__(self, config: DebugConfig):
        self.config = config
        self.traces: list[SampleTrace] = []
        self._current: SampleTrace | None = None
        self._total_seen = 0  # global counter across all batches

    # ── called by FeatureDag.execute() ──

    def begin_sample(self, _batch_sample_id: int = 0) -> None:
        if self._total_seen >= self.config.max_trace_samples:
            self._current = None
            self._total_seen += 1
            return
        self._current = SampleTrace(self._total_seen)
        self._total_seen += 1

    def end_sample(self) -> None:
        if self._current is not None:
            self.traces.append(self._current)
            self._current = None

    def trace_defaults(self, context: dict[str, Any]) -> None:
        if self._current is None:
            return
        out = {n: ValueSnapshot.of(v) for n, v in context.items()}
        self._current.stages.append(StageTrace(StageType.DEFAULT_INIT, "", outputs=out))

    def trace_overrides(self, context: dict[str, Any], overridden: list[str]) -> None:
        if self._current is None:
            return
        inp = {n: ValueSnapshot.of(v) for n, v in context.items() if n not in overridden}
        out = {n: ValueSnapshot.of(v) for n, v in context.items()}
        self._current.stages.append(
            StageTrace(StageType.RAW_OVERRIDE, "", inputs=inp, outputs=out, overridden=overridden)
        )

    def trace_operator(
        self,
        op_name: str,
        input_names: list[str],
        input_vals: list[Any],
        output_names: list[str],
        output_val: Any,
    ) -> None:
        """Record one operator execution."""
        if self._current is None:
            return
        inp = {n: ValueSnapshot.of(v) for n, v in zip(input_names, input_vals)}
        out = {n: ValueSnapshot.of(output_val) for n in output_names}
        stage = StageTrace(StageType.OPERATOR, op_name, inputs=inp, outputs=out)
        # Anomaly detection
        for n, vs in inp.items():
            self._check_anomalies(n, vs, op_name, stage.anomalies)
        for n, vs in out.items():
            self._check_anomalies(n, vs, op_name, stage.anomalies)
        self._current.stages.append(stage)

    # ── anomaly detection ──

    def _check_anomalies(self, name: str, vs: ValueSnapshot, op: str, out: list[Anomaly]) -> None:
        v = vs.value
        if isinstance(v, float) and (math.isnan(v) or math.isinf(v)):
            out.append(Anomaly(name, "NaN/Inf", v))
        elif isinstance(v, list) and len(v) == 0:
            out.append(Anomaly(name, "empty_list"))

    # ── summary / output ──

    def build_summary(self) -> dict:
        per_feature: dict[str, dict] = {}
        per_operator: dict[str, dict] = {}
        total = len(self.traces)
        for t in self.traces:
            for stage in t.stages:
                op = stage.stage_name or stage.stage_type.value
                e = per_operator.setdefault(op, {"executions": 0, "anomalies": 0})
                e["executions"] += 1
                e["anomalies"] += len(stage.anomalies)
                for name, vs in stage.outputs.items():
                    f = per_feature.setdefault(name, {"missing": 0, "anomalies": 0})
                    if (
                        vs.value is None
                        or vs.value == ""
                        or (isinstance(vs.value, float) and math.isnan(vs.value))
                    ):
                        f["missing"] += 1
                    f["anomalies"] += sum(1 for a in stage.anomalies if a.feature == name)
        return {"total_samples": total, "per_feature": per_feature, "per_operator": per_operator}

    def save(self) -> None:
        if not self.config.output_dir:
            return
        Path(self.config.output_dir).mkdir(parents=True, exist_ok=True)
        ts = int(time.time() * 1_000_000)
        # Summary
        summary = self.build_summary()
        sp = Path(self.config.output_dir) / f"summary_{ts}.json"
        with open(sp, "w", encoding="utf-8") as f:
            json.dump(summary, f, indent=2, default=str)
        # Traces (JSONL)
        tp = Path(self.config.output_dir) / f"traces_{ts}.jsonl"
        with open(tp, "w", encoding="utf-8") as f:
            for t in self.traces:
                f.write(json.dumps(t.to_dict(), default=str) + "\n")
        print(f"[Debug] summary → {sp}")
        print(f"[Debug] traces → {tp} ({len(self.traces)} samples)")


@dataclass
class SampleTrace:
    sample_id: int
    stages: list[StageTrace] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {"sample": self.sample_id, "stages": [s.to_dict() for s in self.stages]}
